### PR TITLE
fix(node): duplicated event log when node restarted

### PR DESCRIPTION
**Pull request by mantissahz**
Create the event log for node restart only when `UpdateNodeStatus` without then `conflict` error

longhorn/longhorn#1449

### CODE REVIEW

File: controller/node_controller.go [hunk 1]
```diff
@@ -313,16 +321,17 @@ func (nc *NodeController) syncNode(key string) (err error) {
 			for _, podCondition := range podConditions {
 				if podCondition.Type == v1.PodReady {
 					if podCondition.Status == v1.ConditionTrue && pod.Status.Phase == v1.PodRunning {
-						node.Status.Conditions = types.SetConditionAndRecord(node.Status.Conditions,
-							longhorn.NodeConditionTypeReady, longhorn.ConditionStatusTrue,
-							"", fmt.Sprintf("Node %v is ready", node.Name),
-							nc.eventRecorder, node, v1.EventTypeNormal)
+						// create the event after `UpdateNodeStatus` without the error
+						nodeIsReadyEventMsg = fmt.Sprintf("Node %v is ready", node.Name)
+						nodeIsReadyEventType = v1.EventTypeNormal
```
Comment (shuo-wu) [inline controller/node_controller.go:326]: NIT: You can set Msg and Type only when there is a change. Then you don't need `nodeIsReadyEventChanged`

Comment (mantissahz) [reply controller/node_controller.go]: Modified. Removed `nodeIsReadyEventChanged`. Thanks.

File: controller/node_controller.go [hunk 2]
```diff
@@ -297,6 +299,11 @@ func (nc *NodeController) syncNode(key string) (err error) {
 			logrus.Debugf("Requeue %v due to conflict: %v", key, err)
 			nc.enqueueNode(node)
 			err = nil
+		} else {
+			if err == nil && nodeIsReadyEventMsg != "" && nodeIsReadyEventType != "" {
+				// create the event when the status type `Ready` of node is changed
+				nc.eventRecorder.Event(node, nodeIsReadyEventType, longhorn.NodeConditionTypeReady, nodeIsReadyEventMsg)
+			}
```
Comment (shuo-wu) [inline controller/node_controller.go:306]: NIT: you can refactor this defer function a little bit:
```
if err == nil {
	if !reflect.DeepEqual(existingNode.Status, node.Status) {
		_, err = nc.ds.UpdateNodeStatus(node)
	}
	if nodeIsReadyEventMsg != "" && nodeIsReadyEventType != "" {
		nc.eventRecorder.Event(node, nodeIsReadyEventType, longhorn.NodeConditionTypeReady, nodeIsReadyEventMsg)
	}
} else {
	if apierrors.IsConflict(errors.Cause(err)) {
		logrus.Debugf("Requeue %v due to conflict: %v", key, err)
		nc.enqueueNode(node)
		err = nil
	}
}
```

**Review by shuo-wu — COMMENTED** [review]
BTW. there is a typo in the commit & PR title.

**Review by shuo-wu — DISMISSED** [review]

**Comment by mantissahz** [comment]
>BTW. there is a typo in the commit & PR title.

Modified the typo of the title. Thanks.

**Review by shuo-wu — DISMISSED** [review]

File: controller/node_controller.go [hunk 3]
```diff
@@ -286,17 +286,26 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
-			err = nil
+		if err != nil {
+			// requeue if it's conflict
+			if apierrors.IsConflict(errors.Cause(err)) {
+				logrus.Debugf("Requeue %v due to conflict: %v", key, err)
+				nc.enqueueNode(node)
+				err = nil
+			}
+		} else {
+			if nodeIsReadyEventMsg != "" && nodeIsReadyEventType != "" {
```
Comment (c3y1huang) [inline controller/node_controller.go:305]:
```suggestion
		} else if nodeIsReadyEventMsg != "" && nodeIsReadyEventType != "" {
```

File: controller/node_controller.go [hunk 4]
```diff
@@ -286,17 +286,26 @@ func (nc *NodeController) syncNode(key string) (err error) {
...
-			logrus.Debugf("Requeue %v due to conflict: %v", key, err)
-			nc.enqueueNode(node)
-			err = nil
+		if err != nil {
+			// requeue if it's conflict
+			if apierrors.IsConflict(errors.Cause(err)) {
+				logrus.Debugf("Requeue %v due to conflict: %v", key, err)
+				nc.enqueueNode(node)
+				err = nil
+			}
```
Comment (c3y1huang) [inline controller/node_controller.go:303]: Should we log the error when it is not a conflict error?

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: controller/node_controller.go (hunk 4 above)
Comment (mantissahz) [reply controller/node_controller.go:305]: Yes, added log to record other errors.

File: controller/node_controller.go (hunk 3 above)
Comment (mantissahz) [reply controller/node_controller.go]: Modified. Thanks.

**Review by c3y1huang — APPROVED** [review]

**Review by innobead — APPROVED** [review]

**Pull request merged**